autovalet_lane_detection/scripts/autovalet_lane_detector.py

```diff
@@ -0,0 +1,186 @@
+
+class LaneDetector:
+
+    def __init__(self, color_topic, depth_topic, colorInfo_topic, laneCloud_topic):
+        self.bridge    = CvBridge()
+        self.color_sub = Subscriber(color_topic, Image)
+        self.depth_sub = Subscriber(depth_topic, Image)
+
+        # read camera info by looking up only one message
+        self.camera         = rospy.wait_for_message(colorInfo_topic, CameraInfo)
+        self.laneCloud_pub  = rospy.Publisher(laneCloud_topic, PointCloud2, queue_size=1)
+
+        # Looking to synchronize both the topics within a 1/10th of a second
+        self.ats = ApproximateTimeSynchronizer([self.color_sub, self.depth_sub], queue_size=2, slop=0.5)
+        self.ats.registerCallback(self.image_callback)
+
+        # Define the camera matrix individual values for projection
+        self.fx = self.camera.K[0]
+        self.fy = self.camera.K[4]
+        self.cx = self.camera.K[2]
+        self.cy = self.camera.K[5]
+
+        # define ROI
+        self.ROI_UPPER_Y = 340
+        self.ROI_RIGHT_X = 480
+
+        # tracker
+        self.tracker = None
+
+    def image_callback(self, color_msg, depth_msg):
+        # cvBridge image
+        color_img = self.bridge.imgmsg_to_cv2(color_msg, desired_encoding="bgr8")
+        depth_img = self.bridge.imgmsg_to_cv2(depth_msg, desired_encoding="passthrough")
+
+        # lane detection algo
+        center_line_coordinates = self.center_line_detection(color_img)                   # px2
+        if center_line_coordinates.shape[0] != 0:
+            center_line_cloud       = self.line2cloud(depth_img, center_line_coordinates) # px3
+            lane_cloud              = self.interpolateAnotherLine(center_line_cloud)      # 2px3
+            self.publishLaneCloud(lane_cloud)
+
+    def center_line_detection(self, img):
+        # colorspace transformation
+        rgb = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+        hls = cv2.cvtColor(img, cv2.COLOR_BGR2HLS).astype(np.float)
+
+        # hls thresholding for yellow
+        lower = np.array([12.0, 0.18*255, 0.90*255], dtype=np.uint8)
+        upper = np.array([30.0, 1.00*255, 1.00*255], dtype=np.uint8)
+        mask  = cv2.inRange(hls, lower, upper)
+        th    = cv2.bitwise_and(rgb, rgb, mask=mask).astype(np.uint8)
+        th    = cv2.cvtColor(th, cv2.COLOR_HLS2RGB)
+        th    = cv2.cvtColor(th, cv2.COLOR_RGB2GRAY)
+        th    = cv2.GaussianBlur(th, (5, 5), 0)
+        _, th = cv2.threshold(th, 60, 255, cv2.THRESH_BINARY)
+	th    = cv2.morphologyEx(th, cv2.MORPH_OPEN, (5,5))
+
+        # extract ROI
+        mask = np.zeros_like(th)
+        mask[self.ROI_UPPER_Y:, :self.ROI_RIGHT_X] = 1
+        roi = cv2.bitwise_and(th, th, mask=mask)
+
+        # detect center line
+        minLineLength, maxLineGap = 130, 20
+        edges = cv2.Canny(roi, 100, 200, apertureSize=3)
+        lines = cv2.HoughLinesP(edges, 1, np.pi/180, 30, np.array([]), minLineLength, maxLineGap)
+
+        # possible scenarios: 1. new line 2. tracking
```

Comment:
what about 0?

Reply:
ok yea it's the default value when there are no lines detected. I'll comment more.

Reply:
comment a bit. see if that's enough.